File: AnimalDetails/app/src/main/java/com/example/animaldetails/MainActivity.java
package com.example.animaldetails;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.widget.Toast;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity implements ItemClickListener {

    private ArrayList<Animal> animalList = new ArrayList<>();
    private RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView=findViewById(R.id.recyclerView);
        buildList();
        setRecyclerView();
    }

    private void buildList() {
        for(int i=0; i<20; i++){
            Animal animalDog = new Animal(R.drawable.dog,"dog","bow bow");
            Animal animalCat = new Animal(R.drawable.cat,"cat","meow meow");
            animalList.add(animalDog);
            animalList.add(animalCat);
        }
    }

    private void setRecyclerView() {
        AnimalAdapter animalAdapter = new AnimalAdapter(animalList,this);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(this,2);
        recyclerView.setAdapter(animalAdapter);
        recyclerView.setLayoutManager(gridLayoutManager);
    }

    @Override
    public void onClicked(Animal animal, int position) {
        Toast.makeText(this, animal.getType()+" no. "+position/2, Toast.LENGTH_SHORT).show();
    }
}

File: itunesApp/app/src/main/java/com/example/itunesapp/MainActivity.kt
package com.example.itunesapp

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import androidx.activity.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.GridLayoutManager
import com.example.itunesapp.databinding.ActivityMainBinding
import com.example.itunesapp.ui.MusicAdapter
import com.example.itunesapp.viewModel.MusicViewModel
import dagger.hilt.android.AndroidEntryPoint
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import com.example.itunesapp.data.model.MusicTable


@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    private lateinit var binding:ActivityMainBinding
    private  var list = arrayListOf<MusicTable>()
    private lateinit var adapter:MusicAdapter

    private val viewModel:MusicViewModel by viewModels()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if(checkForInternet(this)){
            binding.btnSearch.setOnClickListener {
                val string = binding.etTerm.text.toString()
//            loadData(string)
//                CoroutineScope(Dispatchers.IO).launch {
//                }
                viewModel.deleteAll()
                viewModel.saveAllMusic(string)

                viewModel.getAllMusics().observe(this, Observer {
                    list.clear()
                    list.addAll(it)
                    adapter.notifyDataSetChanged()
                })

                adapter = MusicAdapter(list)
                val gridLayoutManager = GridLayoutManager(this,2)
                binding.apply {
                    rcRecyclerView.adapter = adapter
                    rcRecyclerView.layoutManager = gridLayoutManager
                }

            }
        }
        else{
            viewModel.getAllMusics().observe(this, Observer {
                list.clear()
                list.addAll(it)
                adapter.notifyDataSetChanged()
            })

            adapter = MusicAdapter(list)
            val gridLayoutManager = GridLayoutManager(this,2)
            binding.apply {
                rcRecyclerView.adapter = adapter
                rcRecyclerView.layoutManager = gridLayoutManager
            }

            binding.btnSearch.setOnClickListener {
                Toast.makeText(this,"Turn On the Internet",Toast.LENGTH_SHORT).show()
            }
        }


    }

//    fun loadData(string:String){
//        viewModel.getAllMusicFromApi(string).observe(this, Observer {
//            when(it.status){
//                Status.ERROR ->{
//                    Toast.makeText(this,"ERROR",Toast.LENGTH_SHORT).show()
//                }
//                Status.LOADING -> {
//                    Toast.makeText(this, "Loading", Toast.LENGTH_SHORT).show()
//                }
//                Status.SUCCESS ->{
//                    list.addAll(it?.data?.results!!)
//                    adapter = MusicAdapter(list)
//                    val gridLayoutManager = GridLayoutManager(this,2)
//                    binding.apply {
//                        rcRecyclerView.adapter = adapter
//                        rcRecyclerView.layoutManager = gridLayoutManager
//                    }
//                    adapter.notifyDataSetChanged()
//                }
//            }
//        })
//    }

    private fun checkForInternet(context: Context): Boolean {

        // register activity with the connectivity manager service
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

        // if the android version is equal to M
        // or greater we need to use the
        // NetworkCapabilities to check what type of
        // network has the internet connection
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {

            // Returns a Network object corresponding to
            // the currently active default data network.
            val network = connectivityManager.activeNetwork ?: return false

            // Representation of the capabilities of an active network.
            val activeNetwork = connectivityManager.getNetworkCapabilities(network) ?: return false

            return when {
                // Indicates this network uses a Wi-Fi transport,
                // or WiFi has network connectivity
                activeNetwork.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> true

                // Indicates this network uses a Cellular transport. or
                // Cellular has network connectivity
                activeNetwork.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> true

                // else return false
                else -> false
            }
        } else {
            // if the android version is below M
            @Suppress("DEPRECATION") val networkInfo =
                connectivityManager.activeNetworkInfo ?: return false
            @Suppress("DEPRECATION")
            return networkInfo.isConnected
        }
    }
}